be/src/vec/aggregate_functions/aggregate_function_window_funnel.h

```diff
@@ -0,0 +1,210 @@
+
+                if (timestamp <= last_timestamp) {
+                    events_timestamp[event_idx] = first_timestamp;
+                    if (event_idx + 1 == max_event_level) {
+                        // Usually, max event level is small.
+                        return max_event_level;
+                    }
+                }
+            }
+        }
+
+        for (int64_t i = events_timestamp.size() - 1; i >= 0; i--) {
+            if (events_timestamp[i].has_value()) {
+                return i + 1;
+            }
+        }
+
+        return 0;
+    }
+
+    void merge(const WindowFunnelState& other) {
+        if (other.events.empty()) {
+            return;
+        }
+
+        int64_t orig_size = events.size();
+        events.insert(std::end(events), std::begin(other.events), std::end(other.events));
+        const auto begin = std::begin(events);
+        const auto middle = std::next(events.begin(), orig_size);
+        const auto end = std::end(events);
+        if (!other.sorted) {
+            std::stable_sort(middle, end);
+        }
+
+        if (!sorted) {
+            std::stable_sort(begin, middle);
+        }
+        std::inplace_merge(begin, middle, end);
+        max_event_level = max_event_level > 0 ? max_event_level : other.max_event_level;
+        window = window > 0 ? window : other.window;
+
+        sorted = true;
+    }
+
+    void write(BufferWritable &out) const {
+        write_var_int(max_event_level, out);
+        write_var_int(window, out);
+        write_var_int(events.size(), out);
+
+        for (int64_t i = 0; i < events.size(); i++) {
+            int64_t timestamp = events[i].first;
+            int event_idx = events[i].second;
+            write_var_int(timestamp, out);
+            write_var_int(event_idx, out);
+        }
+    }
+
+    void read(BufferReadable& in) {
+        int64_t event_level;
+        read_var_int(event_level, in);
+        max_event_level = (int)event_level;
+        read_var_int(window, in);
+        int64_t size = 0;
+        read_var_int(size, in);
+        for (int64_t i = 0; i < size; i++) {
+            int64_t timestamp;
+            int64_t event_idx;
+
+            read_var_int(timestamp, in);
+            read_var_int(event_idx, in);
+            VecDateTimeValue time_value(timestamp);
+            add(time_value, (int)event_idx, max_event_level, window);
+        }
+    }
+};
+
+class AggregateFunctionWindowFunnel
+        : public IAggregateFunctionDataHelper<WindowFunnelState,
+                                              AggregateFunctionWindowFunnel> {
+public:
+    AggregateFunctionWindowFunnel(const DataTypes& argument_types_)
+            : IAggregateFunctionDataHelper<WindowFunnelState,
+                                           AggregateFunctionWindowFunnel>(argument_types_, {}) {
+    }
+
+    String get_name() const override { return "window_funnel"; }
+
+    bool insert_to_null_default() const override { return false; }
```

Comment:
this **return false** means in the process of calculating and get final result, you calculate an illegal value and want to insert null by yourself；

Others if **return true**, it will add null by AggregateFunctionNullBase when input is null

Reply:
I expect window_funnel always return non null value, so i put it in AggregateFunction.NOT_NULLABLE_AGGREGATE_FUNCTION_NAME_SET. Does it make sense?

Reply:
if you are sure it never return null, you should put it in that `_SET`,
and it's may be better to return true about  `insert_to_null_default`;
then it's never return null,  I'm  confused  about `get_return_type` function , you add make_nullable about return type